File: packages/panel-core/src/computeSlots.test.ts
import { describe, it, expect } from 'vitest';
import { computeSlots } from './computeSlots.js';
import type { ElectricalBreaker } from './types.js';

const panel = { total_slots: 8 };

function makeBreaker(overrides: Partial<ElectricalBreaker> & { slot: number }): ElectricalBreaker {
  return {
    id: `b${overrides.slot}`,
    panel_id: 'p1',
    label: null,
    amps: 20,
    breaker_type: 'standard',
    is_gfci: false,
    is_afci: false,
    notes: null,
    floor_plan_area_id: null,
    created_at: '',
    updated_at: '',
    ...overrides,
  };
}

describe('computeSlots — simple panel', () => {
  it('assigns odd slots to left, even to right', () => {
    const slots = computeSlots(panel, []);
    const odd = slots.filter((s) => s.slot % 2 === 1);
    const even = slots.filter((s) => s.slot % 2 === 0);
    expect(odd.every((s) => s.col === 'left')).toBe(true);
    expect(even.every((s) => s.col === 'right')).toBe(true);
  });

  it('produces total_slots entries', () => {
    expect(computeSlots(panel, []).length).toBe(8);
  });

  it('marks all unoccupied slots as blank', () => {
    const slots = computeSlots(panel, []);
    expect(slots.every((s) => s.state === 'blank')).toBe(true);
  });

  it('marks occupied slot correctly', () => {
    const b = makeBreaker({ slot: 3, label: 'Kitchen', breaker_type: 'standard' });
    const slots = computeSlots(panel, [b]);
    const slot3 = slots.find((s) => s.slot === 3)!;
    expect(slot3.state).toBe('occupied');
    expect(slot3.col).toBe('left');
    expect(slot3.isUnlabeled).toBe(false);
  });

  it('marks breaker with no label as isUnlabeled', () => {
    const b = makeBreaker({ slot: 1 });
    const slots = computeSlots(panel, [b]);
    expect(slots.find((s) => s.slot === 1)!.isUnlabeled).toBe(true);
  });
});

describe('computeSlots — double-pole', () => {
  it('renders double-pole as left-column double-height', () => {
    const b = makeBreaker({ slot: 3, breaker_type: 'double_pole', label: 'HVAC' });
    const slots = computeSlots(panel, [b]);
    const dp = slots.find((s) => s.slot === 3)!;
    expect(dp.col).toBe('left');
    expect(dp.rowSpan).toBe(2);
    expect(dp.state).toBe('occupied');
  });

  it('removes the secondary same-column slot and keeps the adjacent column slots', () => {
    const b = makeBreaker({ slot: 3, breaker_type: 'double_pole', label: 'HVAC' });
    const slots = computeSlots(panel, [b]);
    // slot 5 (same column, next row) is consumed
    expect(slots.find((s) => s.slot === 5)).toBeUndefined();
    // slot 4 and slot 6 (opposite column) remain available
    expect(slots.find((s) => s.slot === 4)).toBeDefined();
    expect(slots.find((s) => s.slot === 6)).toBeDefined();
  });

  it('total output length is total_slots - 1 for one double-pole', () => {
    const b = makeBreaker({ slot: 3, breaker_type: 'double_pole' });
    expect(computeSlots(panel, [b]).length).toBe(7);
  });
});

describe('computeSlots — main breaker', () => {
  it('renders main breaker first with full col and row 1', () => {
    const main = makeBreaker({ slot: 0, breaker_type: 'main', label: 'Main 200A' });
    const slots = computeSlots({ total_slots: 4 }, [main]);
    expect(slots[0].state).toBe('main');
    expect(slots[0].col).toBe('full');
    expect(slots[0].row).toBe(1);
  });

  it('regular slots start at row 2 when main is present', () => {
    const main = makeBreaker({ slot: 0, breaker_type: 'main', label: 'Main' });
    const slots = computeSlots({ total_slots: 2 }, [main]);
    const regular = slots.filter((s) => s.state !== 'main');
    expect(regular.every((s) => s.row >= 2)).toBe(true);
  });
});

describe('computeSlots — mixed panel', () => {
  it('handles a realistic panel: main + standard + double-pole + blanks', () => {
    const breakers = [
      makeBreaker({ slot: 0, breaker_type: 'main', label: 'Main 200A' }),
      makeBreaker({ slot: 1, breaker_type: 'standard', label: 'Kitchen', amps: 20 }),
      makeBreaker({ slot: 3, breaker_type: 'double_pole', label: 'HVAC', amps: 60 }),
    ];
    const slots = computeSlots({ total_slots: 8 }, breakers);
    const main = slots.find((s) => s.state === 'main')!;
    const kitchen = slots.find((s) => s.slot === 1)!;
    const hvac = slots.find((s) => s.slot === 3)!;

    expect(main).toBeDefined();
    expect(kitchen.col).toBe('left');
    expect(hvac.col).toBe('left');
    expect(hvac.rowSpan).toBe(2);
    // slot 4 (right column) remains available; slot 5 (same column) is consumed
    expect(slots.find((s) => s.slot === 4)).toBeDefined();
    expect(slots.find((s) => s.slot === 5)).toBeUndefined();
  });
});

File: apps/android/app/src/test/java/com/domitara/domain/ComputeSlotsTest.kt
package com.domitara.domain

import com.domitara.data.dto.BreakerType
import com.domitara.data.dto.ElectricalBreaker
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ComputeSlotsTest {

    private fun breaker(
        slot: Int,
        type: BreakerType = BreakerType.STANDARD,
        label: String? = "L$slot",
    ) = ElectricalBreaker(
        id = "b$slot",
        panelId = "p",
        slot = slot,
        label = label,
        amps = 20,
        breakerType = type,
        isGfci = false,
        isAfci = false,
        notes = null,
        floorPlanAreaId = null,
        createdAt = "",
        updatedAt = "",
    )

    @Test
    fun defaultSlots_knownAndUnknownAmps() {
        assertEquals(20, defaultSlotsForAmps(100))
        assertEquals(30, defaultSlotsForAmps(150))
        assertEquals(40, defaultSlotsForAmps(200))
        assertEquals(84, defaultSlotsForAmps(400))
        assertEquals(20, defaultSlotsForAmps(125)) // fallback
    }

    @Test
    fun blankPanel_allBlanks_correctColsAndRows() {
        val slots = computeSlots(4, emptyList())
        assertEquals(4, slots.size)
        // slot 1 -> left/row1, 2 -> right/row1, 3 -> left/row2, 4 -> right/row2
        val s1 = slots.first { it.slot == 1 }
        val s2 = slots.first { it.slot == 2 }
        val s3 = slots.first { it.slot == 3 }
        val s4 = slots.first { it.slot == 4 }
        assertEquals(SlotCol.LEFT, s1.col); assertEquals(1, s1.row)
        assertEquals(SlotCol.RIGHT, s2.col); assertEquals(1, s2.row)
        assertEquals(SlotCol.LEFT, s3.col); assertEquals(2, s3.row)
        assertEquals(SlotCol.RIGHT, s4.col); assertEquals(2, s4.row)
        slots.forEach { assertEquals(SlotState.BLANK, it.state); assertNull(it.breaker) }
    }

    @Test
    fun mainBreaker_occupiesRow1FullWidth_othersShiftDown() {
        val main = breaker(1, BreakerType.MAIN)
        val slots = computeSlots(4, listOf(main, breaker(2), breaker(3)))
        val mainGeom = slots.first()
        assertEquals(SlotState.MAIN, mainGeom.state)
        assertEquals(SlotCol.FULL, mainGeom.col)
        assertEquals(1, mainGeom.row)
        // slot 1 is consumed by main; regular grid starts at row 2
        assertTrue(slots.none { it.slot == 1 && it.state != SlotState.MAIN })
        // The main consumes slot 1, so slot 2 (even) sits alone in row 2 and the
        // grid row advances; slot 3 (odd) therefore lands in row 3. This matches
        // the original panel-core algorithm exactly.
        val s2 = slots.first { it.slot == 2 }
        assertEquals(2, s2.row)
        assertEquals(SlotCol.RIGHT, s2.col)
        val s3 = slots.first { it.slot == 3 }
        assertEquals(3, s3.row)
        assertEquals(SlotCol.LEFT, s3.col)
    }

    @Test
    fun doublePole_spansTwoRows_consumesSlotPlusTwo() {
        val dp = breaker(1, BreakerType.DOUBLE_POLE)
        val slots = computeSlots(4, listOf(dp))
        val dpGeom = slots.first { it.slot == 1 }
        assertEquals(2, dpGeom.rowSpan)
        assertEquals(SlotCol.LEFT, dpGeom.col)
        assertEquals(SlotState.OCCUPIED, dpGeom.state)
        // slot 3 (1 + 2) is consumed and not rendered separately
        assertTrue(slots.none { it.slot == 3 })
    }

    @Test
    fun doublePole_onEvenSlot_doesNotCollideWithNextOddSlot() {
        // DP on slot 2 (right column) must not push the following odd slot into the
        // same cell as slot 1. Each left slot keeps its own row.
        val slots = computeSlots(4, listOf(breaker(2, BreakerType.DOUBLE_POLE)))
        val s1 = slots.first { it.slot == 1 }
        val s2 = slots.first { it.slot == 2 }
        val s3 = slots.first { it.slot == 3 }
        assertEquals(SlotCol.LEFT, s1.col); assertEquals(1, s1.row)
        assertEquals(SlotCol.RIGHT, s2.col); assertEquals(1, s2.row); assertEquals(2, s2.rowSpan)
        assertEquals(SlotCol.LEFT, s3.col); assertEquals(2, s3.row)
        // slot 4 is the consumed second pole of the double-pole.
        assertTrue(slots.none { it.slot == 4 })
        // No two rendered slots may share the same column and row.
        val cells = slots.map { it.col to it.row }
        assertEquals(cells.size, cells.toSet().size)
    }

    @Test
    fun doublePole_onLastOddSlot_doesNotOverflowPanel() {
        // DP on the last left slot has no paired slot (slot+2 > totalSlots), so it
        // renders as a single row and never extends past the panel's last row.
        val slots = computeSlots(4, listOf(breaker(3, BreakerType.DOUBLE_POLE)))
        val s3 = slots.first { it.slot == 3 }
        assertEquals(1, s3.rowSpan)
        val maxRow = slots.maxOf { it.row + it.rowSpan - 1 }
        assertEquals(2, maxRow) // 4 slots -> 2 rows, no phantom 3rd row
    }

    @Test
    fun unlabeledBreaker_isFlagged() {
        val slots = computeSlots(2, listOf(breaker(1, label = null), breaker(2, label = "Kitchen")))
        assertTrue(slots.first { it.slot == 1 }.isUnlabeled)
        assertTrue(!slots.first { it.slot == 2 }.isUnlabeled)
    }

    @Test
    fun resultSortedByRowThenCol_mainFirst() {
        val main = breaker(1, BreakerType.MAIN)
        val slots = computeSlots(4, listOf(main, breaker(2), breaker(3), breaker(4)))
        assertEquals(SlotState.MAIN, slots.first().state)
        // Verify non-main entries are ordered by (row, col)
        val rest = slots.drop(1)
        val colOrder = mapOf(SlotCol.LEFT to 0, SlotCol.FULL to 1, SlotCol.RIGHT to 2)
        for (i in 1 until rest.size) {
            val prev = rest[i - 1]; val cur = rest[i]
            val prevKey = prev.row * 10 + (colOrder[prev.col] ?: 0)
            val curKey = cur.row * 10 + (colOrder[cur.col] ?: 0)
            assertTrue("expected sorted order", prevKey <= curKey)
        }
    }
}
